### ISSUE TITLE
Game controller improvments

### ISSUE THREAD

**Issue by AJenbo** [diasurgical/devilutionX#418]
The current game controller code has some minor quirks that should be death with.

# auto-aim
- [x] Targeting enemies should take reachability/walk-path (if not ranged weapon) and player direction into consideration. See https://stackoverflow.com/questions/30368632/calculate-distance-on-a-grid-between-2-points this might already be implemented via path_get_h_cost.
- [x] Monsters that can talk should always be last target
- [x] We do not aim at other players, this should happen if we cast a friendly spell or friendly mode is off
- [x] `FindPath` is expensive on slow devices, especially when path is blocked
- [x] Melee aim isn't correctly finding the rothery closest target, or down prioritizing talkingmonsters.
- [x] Target triggers (stairs) if nothing else is around

# misc
- [x] Automap moves in the opposite direction compared to arrow keys and moves to fast.
- [x] Town portal should be cast in front of the player instead of where the mouse happens to be (same for the firewall with no target monster etc)
- [x] Consider a way to hold position (shift attack, spell aim)
- [x] Prevent opening char/inv sheet, and change stats after death or in menu
- [x] When dialog text is being shown, B will close the char sheet
- [x] Inconsistent behavior for "previous", when in the main menu it's B, when a store its Y
- [x] Player can walk across lava and other edges
- [x] Navigate spell book. Left right to flip page, and maybe updown to change spell

- [x] Consider the feasibility of using L1/R1 as modifier buttons to change X/Y/A/B to F5-F8 (favorite spells) the 4 panels instead of the current implementation that only allows for 2 of the 4 panels and no quick spell switching.

- [x] Current button layout differs from the original PSX (and Xbox conventions also), possibly this would be a better default layout:
B: Primary, accept in menus
Y: Secondary, back in menus
X: Cast spell, Use inventory item, delete hero in menu
A: QuickBook
L1: Heal
R1: Mana
L2: Panel mod
R2: Favorit spell mod

Note X/Y A/B is given as laid out on the Nintendo Switch

# cursor 
- [x] Casting targeted spells like resurrection can hinder auto-aim from working, show the cursor and cancel the spell on move or primary action.

**Comment by glebm** [comment]
Attacking on B instead of A felt cramped on RG300, this is why I swapped them. The current mappings match the Nintendo convention: A to confirm, B to cancel. I guess this is different from the PS/XBox convention.

**Comment by MaxDesiatov** [comment]
Should there be a different issue (or it could be discussed here for now) for in-game controller buttons remapping? Different standard controller schemes could be supplied, that switch depending on the controller model, or maybe a player could switch schemes in-game. The easiest stop-gap solution would be to just specify the mapping in the config file.

**Comment by AJenbo** [comment]
@MaxDesiatov for now this issue is a catch all for what is missing from the controller code plus some ideas on how to improve it. As things gets more conceeat it may be split up.

**Comment by dcworldwide** [comment]
Really great work +1

I've just rebuilt after pulling master (ef20d1a) and have tried to use my generic logitech controller on my macbook pro 2015. Controller input not detected. Is there any setup (i.e. compile settings) required to get controller support working?

**Comment by glebm** [comment]
Controller mappings come from the SDL2 database. If your controller is not in the database, it won't be mapped but you can add it to the community -maintained database

**Comment by AJenbo** [comment]
@dcworldwide please run the game from a terminal and copy the output here so that we can see if it detects the controller.

**Comment by dcworldwide** [comment]
From the `build` directory, I've run `open devilutionx.app` however there is no terminal outputting. I must  be missing something simple?

**Comment by MaxDesiatov** [comment]
@dcworldwide you should run it as `./devilutionx.app/Contents/MacOS/devilutionx` instead to get the output

**Comment by dcworldwide** [comment]
Thanks

Here is the output

./devilutionx.app/Contents/MacOS/devilutionx

```
./devilutionx.app/Contents/MacOS/devilutionx
DUMMY: SFileSetBasePath @ SourceX/storm/storm.cpp:807
DUMMY: DispatchMessageA @ SourceX/miniwin/misc_msg.cpp:680
/Users/me/projects/diablo/devilutionX/SourceX/miniwin/misc_msg.cpp:271:35: runtime error: left shift of negative value -22
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Users/me/projects/diablo/devilutionX/SourceX/miniwin/misc_msg.cpp:271:35 in 
DUMMY: UiProfileGetString @ SourceX/DiabloUI/diabloui.cpp:363
DUMMY: CreateFileA : file: /Users/me/Library/Application Support/diasurgical/devilution/single_0.sv (/Users/me/Library/Application Support/diasurgical/devilution/single_0.sv)
```

Booted game, could not navigate menu with controller, loaded existing game using mouse, could not navigate world with controller

**Comment by glebm** [comment]
Another issue: Short d-pad presses do nothing in the inventory screen.

**Comment by AJenbo** [comment]
> Here is the output

It doesn't detect your controller, but I also do not have any idea on how to debug it furhter :/

**Comment by dcworldwide** [comment]
> > Here is the output
> 
> It doesn't detect your controller, but I also do not have any idea on how to debug it furhter :/

No prob. I can try debugging in code. It will be a good opp for me to try and get more familiar with the code. To make my effort faster, any tips on which files I should code some debug console statements? Also if there is a some guidance on how to log to console that would help too

**Comment by AJenbo** [comment]
@dcworldwide grate, this page has a useful code snippet to test with:
https://wiki.libsdl.org/SDL_GameControllerOpen

**Comment by AJenbo** [comment]
Controllers can go where a mouse cannot. In short, the issue is that the controller doesn't check if the move is valid via `path_solid_pieces`. This allows the player to cut corners when a doorway is only blocked by someone directly in front off you, or even cross lave rivers when they bend:

![image](https://user-images.githubusercontent.com/204594/69585513-31c05280-0fe0-11ea-9aef-56506d61f348.png)

**Comment by AJenbo** [comment]
@dcworldwide the controller code should now detect when a game controller has been made avalible, please test if this helps your situatin.

**Comment by dcworldwide** [comment]
@AJenbo works. in-game detected controller insert. Plays very nicely

**Comment by AJenbo** [comment]
Split the issue up in to a few sub issues:
https://github.com/diasurgical/devilutionX/issues/512
https://github.com/diasurgical/devilutionX/issues/511
https://github.com/diasurgical/devilutionX/issues/510

Most of the remaining items in this issue still need some consideration or is simple to deal with

**Comment by AJenbo** [comment]
Having completed the game a few times with the current controller scheme I don't think hold position is needed, or very feasable on a regular controller.

Hot spells have been implemented via https://github.com/diasurgical/devilutionX/pull/513

**Comment by AJenbo** [comment]
Map scrolling has now been flipped to match arrow keys and also move at variable speed and diagonally: https://github.com/diasurgical/devilutionX/commit/cc42e9abb1d29cbf9894a0c3b2a4d31658b2377f

**Comment by MaxBowie** [comment]
Been enjoying the game (1.4.0) with the controller, but couldn't stand the Nintendo OK\Cancel mapping. Was trying to change it via diablo.ini :

[Controller]
Mapping=030000005e040000ff02000000007801,XInput Controller,platform:Windows,a:b0,b:b1,x:b2,y:b3,back:b6,start:b7,leftstick:b8,rightstick:b9,leftshoulder:b4,rightshoulder:b5,dpup:h0.1,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,leftx:a0,lefty:a1,rightx:a3,righty:a4,lefttrigger:a2,righttrigger:a5,

to

[Controller]
Mapping=030000005e040000ff02000000007801,XInput Controller,platform:Windows,a:b1,b:b0,x:b3,y:b2,back:b6,start:b7,leftstick:b8,rightstick:b9,leftshoulder:b4,rightshoulder:b5,dpup:h0.1,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,leftx:a0,lefty:a1,rightx:a3,righty:a4,lefttrigger:a2,righttrigger:a5,

but it had no effect. Not sure if it is a bug or am I doing something wrong?

Basically I think the menu or the ini option to alter mappings would make things much easier.

**Comment by AJenbo** [comment]
Probably try the chat for help

**Comment by dima-ser** [comment]
> Been enjoying the game (1.4.0 on Windows) with the Xbox One controller, but couldn't stand the Nintendo OK\Cancel mapping. Was trying to change it via diablo.ini :
> 
> [Controller] Mapping=030000005e040000ff02000000007801,XInput Controller,platform:Windows,a:b0,b:b1,x:b2,y:b3,back:b6,start:b7,leftstick:b8,rightstick:b9,leftshoulder:b4,rightshoulder:b5,dpup:h0.1,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,leftx:a0,lefty:a1,rightx:a3,righty:a4,lefttrigger:a2,righttrigger:a5,
> 
> to
> 
> [Controller] Mapping=030000005e040000ff02000000007801,XInput Controller,platform:Windows,a:b1,b:b0,x:b3,y:b2,back:b6,start:b7,leftstick:b8,rightstick:b9,leftshoulder:b4,rightshoulder:b5,dpup:h0.1,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,leftx:a0,lefty:a1,rightx:a3,righty:a4,lefttrigger:a2,righttrigger:a5,
> 
> but it had no effect. Not sure if it is a bug or am I doing something wrong?
> 
> Basically I think the menu or the ini option to alter OK\Cancel (and their battle functions) mappings would make things much easier.

Having the same issue. Just need to swap A and B buttons, using XBox Series Controller and putting the mapping in diablo.ini has no effect. Were you able to get it to work?

**Comment by MaxBowie** [comment]
> 

Try this in the ini:

Mapping=030000005e040000ff02000000007200,*,a:b1,b:b0,x:b3,y:b2,back:b6,guide:b10,start:b7,leftstick:b8,rightstick:b9,leftshoulder:b4,rightshoulder:b5,dpup:h0.1,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a4,righttrigger:a5,platform:Windows

If doesnt work - let me know I will tell how to get the right values.

**Comment by dima-ser** [comment]
Doesn't work, I tried variations of this dozens of times. Thanks for trying
though.

I just started playing with the default configuration and sort of got used
to it. Gonna have to relearn again after I switch to another game :)

**Comment by MaxBowie** [comment]
> 

Okay. putting it here just in case anyone else stumbles upon it. 
I used the gamepad tool WITH the latest SDL2.dll - that is very important, without it I was getting wrong values that didnt work. 
So here is the ultility with the updated SDL2 that should give you the right values to put int to the ini. (None of those are mine, I just combined them). Dont forget to swap buttons first : )
https://www.mediafire.com/file/qyzypezgly0l4he/gamepad-tool.rar/file
I still think the simple switch in the ini or in the options would make things so much easier.

**Comment by AJenbo** [comment]
Your not wrong, but we haven't had time to make the gamepad configurable yet. The mapping is really just for adding gamepad that are not already supported, using it for remapping is a bit of a hack.

**Comment by dima-ser** [comment]
Still didn't work. Not sure what I'm doing wrong, I created a custom
mapping in the tool at some point where I had to press each button, but it
doesn't work. I want to go back to default mapping but the tool has no
option to do that.

**Comment by dima-ser** [comment]
Is the GUID string unique per gamepad? Or it's generated based on your
system? How do I know the GUID i have is correct?

**Comment by dima-ser** [comment]
This is what I got from the tool (I reversed a and b):
Mapping=030000005e040000120b000000007801,XboxCustom3,platform:Windows,a:b1,b:b0,x:b2,y:b3,back:b6,start:b7,leftstick:b8,rightstick:b9,leftshoulder:b4,rightshoulder:b5,dpup:h0.1,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,leftx:a0,lefty:a1,rightx:a3,righty:a4,lefttrigger:a2,righttrigger:a5,

**Comment by AJenbo** [comment]
it is dependent on the version of SDL you are using so you have to make sure that the tool uses the same dll
